File: Soutenance 2 End/Miny-RPG Editor/Modfrm.vb
Public Class Modfrm

    Dim Base As Bitmap
    Dim BasePlusGrid As Bitmap

    Private Start As New Point(0, 0)
    Private [End] As New Point(0, 0)
    Private Selecting As Boolean = False

    Public SelectedPoints As New List(Of Point)
    Public SelectedCells As New List(Of Integer)

    Public Sub DrawGrid(ByVal g As Graphics)

        For i As Integer = 0 To Base.Height / 32
            g.DrawLine(Pens.Blue, 0, i * TILE_WIDTH, Base.Width, i * TILE_WIDTH)
        Next

        For i As Integer = 0 To Base.Width / 32
            g.DrawLine(Pens.Blue, i * TILE_WIDTH, 0, i * TILE_WIDTH, Base.Height)
        Next

        For Each Selected As Point In SelectedPoints
            g.DrawRectangle(Pens.Red, Selected.X * TILE_WIDTH, Selected.Y * TILE_WIDTH, TILE_WIDTH, TILE_WIDTH)
        Next

    End Sub

    Private ReadOnly Property TILES_HEIGHT()
        Get
            Return Math.Floor(Base.Height / TILE_WIDTH)
        End Get
    End Property

    Private ReadOnly Property TILES_WIDTH()
        Get
            Return Math.Floor(Base.Width / TILE_WIDTH)
        End Get
    End Property

    Private Sub Modfrm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        SelectedPoints.Add(New Point(0, 0))
        SelectedCells.Add(0)
        Base = New Bitmap("tiles.png")
        RepaintBox()
        Me.Location = Mainfrm.Location + New Point(30, 20)

    End Sub

    Private Sub RepaintBox()

        BasePlusGrid = New Bitmap(Base.Width + 1, Base.Height + 1)
        Dim g As Graphics = Graphics.FromImage(BasePlusGrid)
        g.DrawImage(Base, 0, 0)
        DrawGrid(g)
        PaintBox.Image = BasePlusGrid

    End Sub

    Private Sub PaintBox_MouseDown(ByVal sender As System.Object, ByVal e As MouseEventArgs) Handles PaintBox.MouseDown

        Start = New Point(Math.Floor(e.X / TILE_WIDTH), Math.Floor(e.Y / TILE_WIDTH))
        Selecting = True

    End Sub

    Private Sub ArrangePoints()

        Dim FirstPoint As New Point(500, 500)
        Dim LastPoint As New Point(-1, -1)

        For Each p As Point In SelectedPoints
            If p.X < FirstPoint.X OrElse p.Y < FirstPoint.Y Then
                FirstPoint = p
            End If
            If p.X > LastPoint.X OrElse p.Y > LastPoint.Y Then
                LastPoint = p
            End If
        Next

        SelectedPoints.Clear()
        SelectedCells.Clear()

        Dim Decalage As Point = FirstPoint
        LastPoint -= FirstPoint
        FirstPoint.X = 0
        FirstPoint.Y = 0

        For i As Integer = FirstPoint.X To LastPoint.X
            For j As Integer = FirstPoint.Y To LastPoint.Y
                SelectedPoints.Add(New Point(i, j))
                SelectedCells.Add((j + Decalage.Y) * Math.Floor(Base.Width / TILE_WIDTH) + (i + Decalage.X))
            Next
        Next

    End Sub

    Private Sub PaintSelection()

        SelectedPoints.Clear()
        SelectedCells.Clear()

        For i As Integer = Start.X To [End].X Step If(Start.X > [End].X, -1, 1)
            For j As Integer = Start.Y To [End].Y Step If(Start.Y > [End].Y, -1, 1)
                If i < TILES_WIDTH AndAlso j < TILES_HEIGHT Then
                    SelectedPoints.Add(New Point(i, j))
                    SelectedCells.Add(j * Math.Floor(Base.Width / TILE_WIDTH) + i)
                End If
            Next
        Next

        RepaintBox()

        ArrangePoints()

    End Sub

    Private Sub PaintBox_MouseMove(ByVal sender As System.Object, ByVal e As MouseEventArgs) Handles PaintBox.MouseMove

        If Selecting Then

            [End] = New Point(Math.Floor(e.X / TILE_WIDTH), Math.Floor(e.Y / TILE_WIDTH))
            PaintSelection()

        End If

    End Sub

    Private Sub PaintBox_MouseUp(ByVal sender As System.Object, ByVal e As MouseEventArgs) Handles PaintBox.MouseUp

        If Selecting Then

            [End] = New Point(Math.Floor(e.X / TILE_WIDTH), Math.Floor(e.Y / TILE_WIDTH))
            PaintSelection()

        End If

        Selecting = False

    End Sub

    Private Sub PaintBox_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles PaintBox.Click

    End Sub
End Class

File: Soutenance 3/MapEditorReborn/MapEditorReborn/Modfrm.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

namespace MapEditorReborn
{
    public partial class Modfrm : Form
    {
        public List<Point> SelectedPoints;
        public List<int> SelectedCells;

        private Bitmap Base;
        private Bitmap basePlusGrid;

        private Point start;
        private Point end;
        private bool selecting;

        private int tilesHeight
        {
            get { return (int)Math.Floor((double)Base.Height / Constants.TileWidth); }
        }

        private int tilesWidth
        {
            get { return (int)Math.Floor((double)Base.Width / Constants.TileWidth); }
        }

        public Modfrm()
        {
            InitializeComponent();
            start = new Point(0, 0);
            end = new Point(0, 0);
            selecting = false;

            SelectedPoints = new List<Point>();
            SelectedCells = new List<int>();
        }

        public void DrawGrid(Graphics g)
        {
            for (int i = 0; i < Base.Height; i++)
                g.DrawLine(Pens.Blue, 0, i * Constants.TileWidth, Base.Width, i * Constants.TileWidth);

            for (int i = 0; i < Base.Width; i++)
                g.DrawLine(Pens.Blue, i * Constants.TileWidth, 0, i * Constants.TileWidth, Base.Height);
            foreach (Point p in SelectedPoints)
            {
                g.DrawRectangle(Pens.Red, p.X * Constants.TileWidth, p.Y * Constants.TileWidth, Constants.TileWidth, Constants.TileWidth);
            }
        }

        private void modfrm_Load(object sender, EventArgs e)
        {
            SelectedPoints.Add(new Point(0, 0));
            SelectedCells.Add(0);
            Base = new Bitmap("tiles.png");
            RepaintBox();
            Location = new Point(100, 100);
        }

        private void PaintSelection()
        {
            SelectedPoints.Clear();
            SelectedCells.Clear();

            for (int i = start.X; i <= end.X; i += (start.X > end.X ? -1 : 1))
            {
                for (int j = start.Y; j <= end.Y; j += (start.Y > end.Y ? -1 : 1))
                {
                    if (i < Constants.TileWidth && j < Constants.TileHeight)
                    {
                        SelectedPoints.Add(new Point(i, j));
                        SelectedCells.Add(j * (int)Math.Floor((double)Base.Width / Constants.TileWidth) + i);
                    }
                }
            }
            RepaintBox();
            ArrangePoints();
        }

        private void RepaintBox()
        {
            basePlusGrid = new Bitmap(Base.Width + 1, Base.Height + 1);
            Graphics g = Graphics.FromImage(basePlusGrid);
            g.DrawImage(Base, 0, 0);
            DrawGrid(g);
            PaintBox.Image = basePlusGrid;
        }

        private void ArrangePoints()
        {
            Point firstPoint = new Point(500, 500);
            Point lastPoint = new Point(-1, -1);

            foreach (Point p in SelectedPoints)
            {
                if (p.X < firstPoint.X || p.Y < firstPoint.Y)
                    firstPoint = p;
                if (p.X > lastPoint.X || p.Y > lastPoint.Y)
                    lastPoint = p;
            }

            SelectedPoints.Clear();
            SelectedCells.Clear();

            Point decalage = firstPoint;
            lastPoint.X -= firstPoint.X;
            lastPoint.Y -= firstPoint.Y;
            firstPoint = new Point(0,0);

            for (int i = firstPoint.X; i <= lastPoint.X; i++)
                for (int j = firstPoint.Y; j <= lastPoint.Y; j++)
                {
                    SelectedPoints.Add(new Point(i, j));
                    SelectedCells.Add((j + decalage.Y) * (int)Math.Floor(Base.Width / (double)Constants.TileWidth) +
                                      (i + decalage.X));
                }
        }

        private void PaintBox_Click(object sender, EventArgs e)
        {

        }

        private void PaintBox_MouseDown(object sender, MouseEventArgs e)
        {
            start = new Point((int)Math.Floor((double)e.X / Constants.TileWidth), (int)Math.Floor((double)e.Y / Constants.TileWidth));
            selecting = true;
        }

        private void PaintBox_MouseMove(object sender, MouseEventArgs e)
        {
            if (selecting)
            {
                end = new Point((int)Math.Floor((double)e.X / Constants.TileWidth),
                            (int)Math.Floor((double)e.Y / Constants.TileWidth));
                PaintSelection();
            }
        }

        private void PaintBox_MouseUp(object sender, MouseEventArgs e)
        {
            if (selecting)
            {
                end = new Point((int)Math.Floor((double)e.X / Constants.TileWidth),
                            (int)Math.Floor((double)e.Y / Constants.TileWidth));
                PaintSelection();
            }
            selecting = false;
        }
    }
}
